pkg/agent/apiserver/handlers/memberlist/handler_test.go

```diff
@@ -68,35 +68,75 @@ func TestMemberlistQuery(t *testing.T) {
 	informerFactory.Start(stopCh)
 	informerFactory.WaitForCacheSync(stopCh)
 
-	ctrl := gomock.NewController(t)
-	q := queriertest.NewMockAgentQuerier(ctrl)
-	memberlistInterface := memberlisttest.NewMockInterface(ctrl)
-	q.EXPECT().GetNodeLister().Return(nodeLister)
-	q.EXPECT().GetMemberlistCluster().Return(memberlistInterface)
-	memberlistInterface.EXPECT().AliveNodes().Return(sets.NewString("node1"))
-	handler := HandleFunc(q)
-
-	req, err := http.NewRequest(http.MethodGet, "", nil)
-	require.NoError(t, err)
-
-	recorder := httptest.NewRecorder()
-	handler.ServeHTTP(recorder, req)
-	assert.Equal(t, http.StatusOK, recorder.Code)
-
-	expectedResponse := []Response{
+	tests := []struct {
+		name                  string
+		mockNodeLister        func(*queriertest.MockAgentQuerier)
+		mockMemberlistCluster func(*queriertest.MockAgentQuerier, *memberlisttest.MockInterface)
+		mockAliveNodes        func(*memberlisttest.MockInterface)
+		expectedStatus        int
+		expectedResponse      []Response
+	}{
 		{
-			NodeName: "node1",
-			IP:       "172.16.0.11",
-			Status:   "Alive",
+			name: "memberlist not running",
+			mockMemberlistCluster: func(q *queriertest.MockAgentQuerier, m *memberlisttest.MockInterface) {
+				q.EXPECT().GetMemberlistCluster().Return(nil)
+			},
+			expectedStatus: http.StatusServiceUnavailable,
 		},
 		{
-			NodeName: "node2",
-			IP:       "172.16.0.12",
-			Status:   "Dead",
+			name: "get memberlist",
+			mockNodeLister: func(q *queriertest.MockAgentQuerier) {
+				q.EXPECT().GetNodeLister().Return(nodeLister)
+			},
+			mockMemberlistCluster: func(q *queriertest.MockAgentQuerier, m *memberlisttest.MockInterface) {
+				q.EXPECT().GetMemberlistCluster().Return(m).Times(2)
+			},
+			mockAliveNodes: func(m *memberlisttest.MockInterface) {
+				m.EXPECT().AliveNodes().Return(sets.NewString("node1"))
+			},
+			expectedStatus: http.StatusOK,
+			expectedResponse: []Response{
+				{
+					NodeName: "node1",
+					IP:       "172.16.0.11",
+					Status:   "Alive",
+				},
+				{
+					NodeName: "node2",
+					IP:       "172.16.0.12",
+					Status:   "Dead",
+				},
+			},
 		},
 	}
-	var received []Response
-	err = json.Unmarshal(recorder.Body.Bytes(), &received)
-	require.NoError(t, err)
-	assert.ElementsMatch(t, expectedResponse, received)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			q := queriertest.NewMockAgentQuerier(ctrl)
+			if tt.expectedStatus == http.StatusOK {
+				memberlistInterface := memberlisttest.NewMockInterface(ctrl)
+				tt.mockNodeLister(q)
+				tt.mockMemberlistCluster(q, memberlistInterface)
+				tt.mockAliveNodes(memberlistInterface)
+			} else {
+				tt.mockMemberlistCluster(q, nil)
+			}
```

Comment:
```suggestion
			q.GetMemberlistCluster().Return(tt.memberlistInterface(ctrl))
			q.GetNodeLister().Return(tt.nodeLister(ctrl))
```
And set up mock interface in the case's own memberlistInterface() and nodeLister().